feat(participants-pane) - ParticipantsPane native component

Calinteodor opened the pull request:
<!--
Thank you for your pull request. Please provide a thorough description below.

Contributors guide: https://github.com/jitsi/jitsi-meet/blob/master/CONTRIBUTING.md
-->

Calinteodor commented on react/features/video-menu/actions.any.js:
```diff
@@ -98,16 +98,33 @@ export function muteAllParticipants(exclude: Array<string>, mediaType: MEDIA_TYP
     };
 }
+
+
```
This is something that I still got to work on.

Calinteodor commented on react/features/video-menu/actions.any.js:
```diff
@@ -98,16 +98,33 @@ export function muteAllParticipants(exclude: Array<string>, mediaType: MEDIA_TYP
+ *
+ * @returns {Function}
+ */
+export function dontAllowUnmute() {
```
This is something in progress.

tmoldovan8x8 commented on react/features/participants-pane/components/native/MeetingParticipantItem.js:
```diff
@@ -0,0 +1,39 @@
+            audioMuteState = { isAudioMuted ? MediaState.Muted : MediaState.Unmuted }
+            isKnockingParticipant = { false }
+            name = { p.name }
+            onPress = { !p.local && openContextMenuDetails }
```
There is a warning thrown I think on this onPress

Calinteodor replied:
Fixed.

tmoldovan8x8 commented on react/features/participants-pane/components/native/ContextMenuMore.js:
```diff
@@ -0,0 +1,60 @@
+            onCancel = { cancel }
+            style = { styles.contextMenuMore }>
+            <TouchableOpacity
+                onPress = { muteEveryoneVideo }
```
This throws an error, does not seem to work

Calinteodor replied:
Fixed.

tmoldovan8x8 commented on react/features/participants-pane/components/native/ContextMenuMore.js:
```diff
@@ -0,0 +1,60 @@
+                <Text style = { styles.contextMenuItemText }>{t('participantsPane.actions.stopEveryonesVideo')}</Text>
+            </TouchableOpacity>
+            <TouchableOpacity
+                onPress = { blockAudioVideo }
```
Also this

Calinteodor replied:
This will be covered with the a-v moderation feature.

tmoldovan8x8 commented on react/features/participants-pane/components/native/MeetingParticipantList.js:
```diff
@@ -0,0 +1,56 @@
+            }
+        </View>
+    );
+};
```
Could you please check the behaviour while screen rotating on Android. I've seen some strange UI on this component, but might be on others, as well

Calinteodor replied:
Can you please attach a screenshot with this because I cannot reproduce it. Thanks.

vp8x8 commented on react/features/participants-pane/components/native/ParticipantItem.js:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * Callback for when the mouse leaves this component
+     */
+    onLeave?: Function,
```
I don't think we need this on mobile.

Calinteodor replied:
Removed.

vp8x8 commented on react/features/participants-pane/components/native/ParticipantItem.js:
```diff
@@ -0,0 +1,116 @@
+    return (
+        <View style = { styles.participantContainer } >
+            <TouchableOpacity
+                /* eslint-disable-next-line react/jsx-no-bind */
```
I see no bind here. Is this comment necessary?

Calinteodor replied:
Removed.

vp8x8 commented on react/features/video-menu/actions.any.js:
```diff
@@ -92,3 +94,40 @@ export function muteAllParticipants(exclude: Array<string>, mediaType: MEDIA_TYP
+        const state = getState();
+        const participants = state['features/base/participants'];
+
+        console.log(participants);
```
I don't think we need this log.

vp8x8 commented on react/features/video-menu/components/native/VolumeSlider.js:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @returns {React$Element<any>}
+ */
+class VolumeSlider extends Component<Props, State> {
```
Any reason for creating a class component instead of a functional one?

Calinteodor replied:
The only reason was that the components created for this feature are class based.

vp8x8 replied:
Could you please use a functional component? The ideea is to try to use functional components wherever possible and hooks. So instead of inheritance we could reuse code through hooks.

damencho replied:
We need to discuss this. I'm not 100% sure that this is a good idea. There can be performance issues and re-rendering for those components, we were debugging the participants' pane the last few days ...

Calinteodor replied:
Updated

vp8x8 replied:
@damencho There should be no difference in performance between class components & functional components. If there are multiple re-renders this must be because of how the component was built. I guess we should use React memo & Pure component more often.

damencho replied:
Yeah, I guess it was a bad implementation of the selectors as there were some places returning new objects (new constructed arrays) which will trigger new re-renders ... I'm not even sure where that was at the moment, we will soon create the PR.  
But there is a difference in how the comparison is done by react in the selectors it is === while in mapstatet2props it is deep equal comparison, so there is a big difference in this which can lead to re-renders.

damencho replied:
An example I just saw https://github.com/jitsi/jitsi-meet/blob/2174368d23e115781118b23ed6f00ce95ab30beb/react/features/av-moderation/functions.js#L95
This is used in a selector, this will be rendered on every action

damencho replied:
And maybe I added that, by the way :) not sure

vp8x8 commented:
This needs a rebase.

Calinteodor commented on react/features/base/react/components/native/HeaderWithNavigation.js:
```diff
@@ -48,17 +53,22 @@ class HeaderWithNavigation extends Component<Props> {
-                    labelKey = { this.props.forwardLabelKey }
-                    onPress = { onPressForward } /> }
-            </Header>
+            <>
```
Fixed.

Calinteodor commented on react/features/conference/components/native/Conference.js:
```diff
@@ -71,6 +72,11 @@ type Props = AbstractProps & {
+    /**
+     * The indicator which determines if the participants pane is open.
+     */
+    _isOpen: boolean,
```
Renamed.

Calinteodor commented on react/features/participants-pane/components/native/ContextMenuMeetingParticipantDetails.js:
```diff
@@ -0,0 +1,178 @@
+export const ContextMenuMeetingParticipantDetails = ({ participant: p }: Props) => {
+    const [ volume, setVolume ] = useState(undefined);
+    const store = useStore();
+    const startSilent = store.getState['features/base/config'];
```
Removed useStore and replaced with useSelector.

Calinteodor commented on react/features/participants-pane/components/native/LobbyParticipantList.js:
```diff
@@ -0,0 +1,56 @@
+    const dispatch = useDispatch();
+    const admitAll = useCallback(() =>
+        dispatch(admitAllKnockingParticipants(participants, lobbyEnabled)),
+        [ dispatch ]);
```
Replaced with admitMultiple action.

Calinteodor commented on react/features/participants-pane/components/native/MeetingParticipantList.js:
```diff
@@ -0,0 +1,56 @@
+                && <Button
+                    children = { t('participantsPane.actions.invite') }
+                    /* eslint-disable-next-line react/jsx-no-bind */
+                    icon = { () =>
```
This is react-native-paper framework specific. 
For other than material icons, I can only pass a function which returns a react element to be used an icon.

Calinteodor commented on react/features/participants-pane/components/native/ContextMenuMeetingParticipantDetails.js:
```diff
@@ -0,0 +1,178 @@
+    const startSilent = store.getState['features/base/config'];
+    const dispatch = useDispatch();
+    const cancel = useCallback(() => dispatch(hideDialog()), [ dispatch ]);
+    const changeVolume = useCallback(() => setVolume(volume), [ volume ]);
```
If I rely only on the setVolume, I enter in an infinite render loop. 
A "Too many re-renders" error is thrown.

vp8x8 commented on react/features/video-menu/components/native/VolumeSlider.js:
```diff
@@ -0,0 +1,164 @@
+    const { startSilent } = state['features/base/config'];
+    const { participantsVolume } = state['features/participants-pane'];
+    const getParticipant = participant.id
+        ? getParticipantById(state, participant.id) : getLocalParticipant(state);
```
Why is this necessary? Don't we already have the `id` of the participant?

Calinteodor replied:
Fixed.

vp8x8 commented on react/features/video-menu/components/native/VolumeSlider.js:
```diff
@@ -0,0 +1,164 @@
+
+    return {
+        _startSilent: Boolean(startSilent),
+        _volume: participant.local ? undefined : participantsVolume[id]
```
You could create an separate selector for this.

vp8x8 commented:
This looks pretty good (functionality wise). A few remarks:

1. If a participant quits while the the participants pane & drawer are open for it, there is an error:
<img width="457" alt="Screenshot 2021-07-02 at 12 11 52" src="https://user-images.githubusercontent.com/37841821/124251621-d137b680-db2e-11eb-967e-dda481fef3b2.png">

2. There are a few design inconsistency:
* The upper right close button looks way to big. Please check the dimensions from the design and use the tokens.
* The same applies for the bottom buttons (Mute and ... button)
* The avatar  from the participant list and the icons from  actions menu looks like are not middle aligned.
* There are a few keys missing: `participantsPane.actions.mute` & `participantsPane.actions.muteEveryoneElse`
* The knob for the slider is way too big. 

3. Can we get rid of the warning regarding the Slider? It looks like it's deprecated and we have to use the one from react-native-community instead of the core one.

Calinteodor commented on react/features/base/react/components/native/HeaderWithNavigation.js:
```diff
@@ -48,17 +53,18 @@ class HeaderWithNavigation extends Component<Props> {
-                    labelKey = { this.props.forwardLabelKey }
-                    onPress = { onPressForward } /> }
-            </Header>
+            !{ hideHeaderWithNavigation }
```
> This looks pretty good (functionality wise). A few remarks:
> 
> 1. If a participant quits while the the participants pane & drawer are open for it, there is an error:
> 
> <img alt="Screenshot 2021-07-02 at 12 11 52" width="457" src="https://user-images.githubusercontent.com/37841821/124251621-d137b680-db2e-11eb-967e-dda481fef3b2.png">
> 
> 1. There are a few design inconsistency:
> 
> * The upper right close button looks way to big. Please check the dimensions from the design and use the tokens.
> * The same applies for the bottom buttons (Mute and ... button)
> * The avatar  from the participant list and the icons from  actions menu looks like are not middle aligned.
> * There are a few keys missing: `participantsPane.actions.mute` & `participantsPane.actions.muteEveryoneElse`
> * The knob for the slider is way too big.
> 
> 1. Can we get rid of the warning regarding the Slider? It looks like it's deprecated and we have to use the one from react-native-community instead of the core one.

Everything is fixed except the knob. For the Slider component we have no thumbSize prop. 
There is an issue on the community that is still open
https://github.com/callstack/react-native-slider/issues/97.

vp8x8 commented:
3 more things:

1. The icons from the BottomSheet look too big. They're definitely bigger than 16px.
2. If a participant quits while BottomSheet is opened it remains opened with the respective actions. It should automatically close when a participant leaves.
3. I think you could test the android build using the community slider if you modify 2 files: `/android/settings.graddle`, and `android/sdk/build.gradle`. Make the adjustments similar to `netinfo` package.

Calinteodor commented:
> 3 more things:
> 
> 1. The icons from the BottomSheet look too big. They're definitely bigger than 20px.
> 2. If a participant quits while BottomSheet is opened it remains opened with the respective actions. It should automatically close when a participant leaves.
> 3. I think you could test the android build using the community slider if you modify 2 files: `/android/settings.graddle`, and `android/sdk/build.gradle`. Make the adjustments similar to `netinfo` package.

Fixed 1 and 2. Thanks for the hint regarding 3, but it doesn't work after I updated those two files. 
Also, please don't merge this because @damencho and @hristoterezov change how the participants are stored and we should wait for their PR to be merged first. Thanks!

hristoterezov commented:
@vp8x8 I believe all change requests were addressed. Please let me know if I missed something so that @Calinteodor can fix it.
